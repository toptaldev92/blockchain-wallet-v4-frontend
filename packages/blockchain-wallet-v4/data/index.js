export { default as walletV1 } from './wallet.v1'
export { default as walletV2 } from './wallet.v2'
export { default as walletV3 } from './wallet.v3'
