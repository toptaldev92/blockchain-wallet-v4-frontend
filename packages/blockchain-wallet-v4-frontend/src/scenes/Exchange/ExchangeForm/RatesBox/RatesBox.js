import { connect } from 'react-redux'
import { FormattedMessage } from 'react-intl'
import { getData } from './selectors'
import { LargeTableRow, Wrapper } from 'components/Exchange'
import { path } from 'ramda'
import { SkeletonRectangle, Text } from 'blockchain-info-components'
import React from 'react'
import StringDisplay from 'components/Display/StringDisplay'
import styled from 'styled-components'

const RatesWrapper = styled(Wrapper)`
  padding: 0;
`
const RateText = styled(Text)`
  font-size: 12px;
`
const TotalBalanceWrapper = styled.div`
  text-align: left;
  > ${RateText} {
    margin-bottom: 2px;
  }
`
const RateWrapper = styled.div`
  text-align: right;
`
const Row = styled.div`
  display: flex;
`

export const RatesBox = ({
  balance,
  sourceToTargetRate,
  sourceCoinTicker,
  targetCoinTicker
}) => (
  <RatesWrapper>
    <LargeTableRow>
      <TotalBalanceWrapper>
        <RateText>
          <FormattedMessage
            id='scenes.exchange.exchangeform.summary.totalbalanceavailable'
            defaultMessage='{sourceCoinTicker} available for Swap'
            values={{ sourceCoinTicker }}
          />
        </RateText>
        <RateText>
          {balance.cata({
            Success: ({ balanceMax, balanceMaxFiat }) => {
              return (
                <Row>
                  <RateText color='success'>{balanceMaxFiat} </RateText>
                  <span>&nbsp;</span>
                  <RateText>
                    {path(['amount'], balanceMax)}{' '}
                    {path(['symbol'], balanceMax)}
                  </RateText>
                </Row>
              )
            },
            Failure: () => null,
            Loading: () => <SkeletonRectangle height='14px' width='60px' />,
            NotAsked: () => <SkeletonRectangle height='14px' width='60px' />
          })}
        </RateText>
      </TotalBalanceWrapper>
      <RateWrapper>
        <RateText color='blue900'>1 {sourceCoinTicker} =</RateText>
        <RateText>
          <StringDisplay>
            {sourceToTargetRate.map(rate => `${rate} ${targetCoinTicker}`)}
          </StringDisplay>
        </RateText>
      </RateWrapper>
    </LargeTableRow>
  </RatesWrapper>
)

export default connect(getData)(RatesBox)
