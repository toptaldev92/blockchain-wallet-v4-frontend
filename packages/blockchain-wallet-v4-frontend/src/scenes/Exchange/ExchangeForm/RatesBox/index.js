import RatesBox from './RatesBox'
export default RatesBox
