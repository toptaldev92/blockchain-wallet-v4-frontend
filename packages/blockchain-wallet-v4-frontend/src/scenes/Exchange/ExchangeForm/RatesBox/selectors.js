import { curry, prop, path, lift } from 'ramda'

import { selectors, model } from 'data'
import { createDeepEqualSelector } from 'services/ReselectHelper'
import { Exchange } from 'blockchain-wallet-v4/src'
import { formatFiat } from 'blockchain-wallet-v4/src/exchange/currency'

const { formatPair } = model.rates

const getCurrentPairRates = (state, { sourceCoin, targetCoin }) =>
  selectors.components.exchange.getRates(
    formatPair(sourceCoin, targetCoin),
    state
  )

const fallbackToBestRates = (adviceRatesR, bestRatesR) =>
  adviceRatesR.cata({
    Success: () => adviceRatesR,
    Failure: () => bestRatesR,
    Loading: () => adviceRatesR,
    NotAsked: () => bestRatesR
  })

const formatBestRates = curry(
  (sourceCoin, targetCoin, currency, bestRates) => ({
    sourceToTargetRate: path(
      [formatPair(sourceCoin, targetCoin), 'price'],
      bestRates
    ),
    sourceToFiatRate: path(
      [formatPair(sourceCoin, currency), 'price'],
      bestRates
    ),
    targetToFiatRate: path(
      [formatPair(targetCoin, currency), 'price'],
      bestRates
    )
  })
)

const transformBalanceMax = curry((currency, limits, rates) => {
  const balanceMax = path([currency, 'balanceMax'], limits)
  const sourceToFiatRate = prop('sourceToFiatRate', rates)
  const symbol = Exchange.getSymbol(currency)
  const maxFiat = formatFiat(prop('amount', balanceMax) * sourceToFiatRate)
  return {
    balanceMax,
    balanceMaxFiat: symbol + maxFiat
  }
})

export const getData = createDeepEqualSelector(
  [
    getCurrentPairRates,
    selectors.modules.rates.getBestRates,
    selectors.components.exchange.getLimits,
    (state, ownProps) =>
      selectors.core.walletOptions
        .getCoinModel(state, ownProps.sourceCoin)
        .getOrElse({}),
    (state, ownProps) =>
      selectors.core.walletOptions
        .getCoinModel(state, ownProps.targetCoin)
        .getOrElse({}),
    (state, ownProps) => ownProps
  ],
  (
    adviceRatesR,
    bestRatesR,
    limitsR,
    sourceCoinModel,
    targetCoinModel,
    { currency, sourceCoin, targetCoin }
  ) => {
    const ratesR = fallbackToBestRates(
      adviceRatesR,
      bestRatesR.map(formatBestRates(sourceCoin, targetCoin, currency))
    )
    return {
      balance: lift(transformBalanceMax(currency))(limitsR, ratesR),
      sourceToTargetRate: ratesR.map(prop('sourceToTargetRate')),
      sourceCoinTicker: prop('coinTicker', sourceCoinModel),
      targetCoinTicker: prop('coinTicker', targetCoinModel)
    }
  }
)
