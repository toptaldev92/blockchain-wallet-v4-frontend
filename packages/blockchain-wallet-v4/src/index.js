import * as Coin from './coinSelection/coin'
import * as coreActions from './redux/actions'
import * as coreActionsTypes from './redux/actionTypes'
import * as coreMiddleware from './redux/middleware'
import * as coreReducers from './redux/reducers'
import * as coreSelectors from './redux/selectors'
import * as crypto from './walletCrypto'
import * as Exchange from './exchange'
import * as Network from './network/index.ts'
import * as pairing from './pairing'
import * as paths from './redux/paths'
import * as transactions from './transactions'
import * as Types from './types'
import * as utils from './utils'
import coreRootSagaFactory from './redux/rootSaga'
import coreSagasFactory from './redux/sagas'
import Remote from './remote'

export {
  coreActions,
  coreActionsTypes,
  coreMiddleware,
  coreReducers,
  coreSelectors,
  coreSagasFactory,
  coreRootSagaFactory,
  paths,
  Network,
  Coin,
  Exchange,
  crypto,
  pairing,
  transactions,
  Types,
  Remote,
  utils
}
