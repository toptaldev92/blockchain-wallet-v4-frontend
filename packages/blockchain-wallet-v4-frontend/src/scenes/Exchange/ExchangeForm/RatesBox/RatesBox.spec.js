import React from 'react'
import { forEachObjIndexed } from 'ramda'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { Remote } from 'blockchain-wallet-v4'
import { RatesBox } from './RatesBox'

const props = {
  sourceCoin: 'BTC',
  targetCoin: 'ETH',
  currency: '$',
  balance: Remote.of({ balanceMax: {}, balanceMaxFiat: '$1000.00' })
}

const STUB_RATE = 100

const rateValues = {
  notAsked: Remote.NotAsked,
  loading: Remote.Loading,
  success: Remote.Success(STUB_RATE),
  failure: Remote.Failure({})
}

describe('Exchange RatesBox', () => {
  forEachObjIndexed(
    sourceToTargetRate =>
      it('should render correct rates for sourceToTarget', () => {
        const component = shallow(
          <RatesBox {...props} sourceToTargetRate={sourceToTargetRate} />
        )

        const tree = toJson(component)
        expect(tree).toMatchSnapshot()
      }),
    rateValues
  )
})
