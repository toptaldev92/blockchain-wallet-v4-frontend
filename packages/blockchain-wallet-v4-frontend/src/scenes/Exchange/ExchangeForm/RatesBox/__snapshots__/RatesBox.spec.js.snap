// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange RatesBox should render correct rates for sourceToTarget 1`] = `
<Styled(styled.div)>
  <Styled(styled.div)>
    <styled.div>
      <Styled(Text)>
        <FormattedMessage
          defaultMessage="{sourceCoinTicker} available for Swap"
          id="scenes.exchange.exchangeform.summary.totalbalanceavailable"
          values={
            Object {
              "sourceCoinTicker": undefined,
            }
          }
        />
      </Styled(Text)>
      <Styled(Text)>
        <styled.div>
          <Styled(Text)
            color="success"
          >
            $1000.00
             
          </Styled(Text)>
          <span>
             
          </span>
          <Styled(Text)>
             
          </Styled(Text)>
        </styled.div>
      </Styled(Text)>
    </styled.div>
    <styled.div>
      <Styled(Text)
        color="blue900"
      >
        1 
         =
      </Styled(Text)>
      <Styled(Text)>
        <StringDisplayContainer
          skeletonHeight="14px"
          skeletonWidth="60px"
        >
          <Component />
        </StringDisplayContainer>
      </Styled(Text)>
    </styled.div>
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`Exchange RatesBox should render correct rates for sourceToTarget 2`] = `
<Styled(styled.div)>
  <Styled(styled.div)>
    <styled.div>
      <Styled(Text)>
        <FormattedMessage
          defaultMessage="{sourceCoinTicker} available for Swap"
          id="scenes.exchange.exchangeform.summary.totalbalanceavailable"
          values={
            Object {
              "sourceCoinTicker": undefined,
            }
          }
        />
      </Styled(Text)>
      <Styled(Text)>
        <styled.div>
          <Styled(Text)
            color="success"
          >
            $1000.00
             
          </Styled(Text)>
          <span>
             
          </span>
          <Styled(Text)>
             
          </Styled(Text)>
        </styled.div>
      </Styled(Text)>
    </styled.div>
    <styled.div>
      <Styled(Text)
        color="blue900"
      >
        1 
         =
      </Styled(Text)>
      <Styled(Text)>
        <StringDisplayContainer
          skeletonHeight="14px"
          skeletonWidth="60px"
        >
          <Component />
        </StringDisplayContainer>
      </Styled(Text)>
    </styled.div>
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`Exchange RatesBox should render correct rates for sourceToTarget 3`] = `
<Styled(styled.div)>
  <Styled(styled.div)>
    <styled.div>
      <Styled(Text)>
        <FormattedMessage
          defaultMessage="{sourceCoinTicker} available for Swap"
          id="scenes.exchange.exchangeform.summary.totalbalanceavailable"
          values={
            Object {
              "sourceCoinTicker": undefined,
            }
          }
        />
      </Styled(Text)>
      <Styled(Text)>
        <styled.div>
          <Styled(Text)
            color="success"
          >
            $1000.00
             
          </Styled(Text)>
          <span>
             
          </span>
          <Styled(Text)>
             
          </Styled(Text)>
        </styled.div>
      </Styled(Text)>
    </styled.div>
    <styled.div>
      <Styled(Text)
        color="blue900"
      >
        1 
         =
      </Styled(Text)>
      <Styled(Text)>
        <StringDisplayContainer
          skeletonHeight="14px"
          skeletonWidth="60px"
        >
          <Component />
        </StringDisplayContainer>
      </Styled(Text)>
    </styled.div>
  </Styled(styled.div)>
</Styled(styled.div)>
`;

exports[`Exchange RatesBox should render correct rates for sourceToTarget 4`] = `
<Styled(styled.div)>
  <Styled(styled.div)>
    <styled.div>
      <Styled(Text)>
        <FormattedMessage
          defaultMessage="{sourceCoinTicker} available for Swap"
          id="scenes.exchange.exchangeform.summary.totalbalanceavailable"
          values={
            Object {
              "sourceCoinTicker": undefined,
            }
          }
        />
      </Styled(Text)>
      <Styled(Text)>
        <styled.div>
          <Styled(Text)
            color="success"
          >
            $1000.00
             
          </Styled(Text)>
          <span>
             
          </span>
          <Styled(Text)>
             
          </Styled(Text)>
        </styled.div>
      </Styled(Text)>
    </styled.div>
    <styled.div>
      <Styled(Text)
        color="blue900"
      >
        1 
         =
      </Styled(Text)>
      <Styled(Text)>
        <StringDisplayContainer
          skeletonHeight="14px"
          skeletonWidth="60px"
        >
          <Component />
        </StringDisplayContainer>
      </Styled(Text)>
    </styled.div>
  </Styled(styled.div)>
</Styled(styled.div)>
`;
